### DISCUSSION TITLE
Match all routes with new middleware matcher

### DISCUSSION THREAD

**Discussion by pickled-cucumber** [vercel/next.js#38615] [Help] [answered]
How can we match all routes in an application with the new middleware matcher?

**Comment by icyJoseph** [comment, marked as answer]
If you don't define any matcher, it matches all routes by default. 

> Middleware will be invoked for every route in the app, and a custom matcher can be used to define matching filters. - https://nextjs.org/docs/messages/middleware-upgrade-guide#how-to-upgrade

### Solution

OP's middleware was rewriting everything, as expected by the middleware and OP's code.

```ts
import { NextResponse, type NextRequest } from "next/server";

const PUBLIC_FILE = /\.(.*)$/;

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;
  console.log("Middleware", { pathname });

  if (
    pathname.startsWith("/_next") || // exclude Next.js internals
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    PUBLIC_FILE.test(pathname) // exclude all files in the public folder
  )
    return NextResponse.next();

  // do something else
  return NextResponse.next();
}
```

**Reply by pickled-cucumber** [reply]
Hmm. I figured that might be the case but I end up getting an error in the browser console when the matcher isn't there. That's probably not for this question/thread. Thank you!

<img width="537" alt="image" src="https://user-images.githubusercontent.com/100007168/178807721-460b2a42-9247-4592-9b00-260c04330b13.png">


```import type { NextRequest } from 'next/server'

import { NextResponse } from 'next/server'

// Block Austria, prefer Germany
const ALLOWED_COUNTRY = 'US'

// Limit middleware pathname config
// export const config = {
//   matcher: ['/', '//:path*'],
// }

export function middleware(req: NextRequest) {
  // Extract country
  const country = req.geo?.country || ''

  // Specify the correct pathname
  if (country === ALLOWED_COUNTRY || process.env.NODE_ENV === 'development') {
    req.nextUrl.pathname = '/'
  } else {
    req.nextUrl.pathname = `/info`
  }

  // Rewrite to URL
  return NextResponse.rewrite(req.nextUrl)
}```

**Reply by icyJoseph** [reply]
That's because the middleware is literally matching everything, even Next internals.

```ts
  if (
    pathname.startsWith("/_next") || // exclude Next.js internals
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    PUBLIC_FILE.test(pathname)  // exclude all files in the public folder
  )
    return NextResponse.next();
```

Put this at the top of your middleware, where PUBLIC_FILE is this regex `const PUBLIC_FILE = /\.(.*)$/;`.

**Reply by pickled-cucumber** [reply]
Ah that did it. Thank you so much! I didn't realize it would cover even the internal paths.

**Reply by er-santosh** [reply]
@pickled-cucumber can you please help me out here. i am using the same way of protecting routes using middleware and it is working fine. 
But how am i supposed to make use of this middleware not only in server side, in client side as well. currently i have to create some auth guard and wrap it to the layout so that it can watch the authentication check. 

`

import { authCookieKey } from "@/libs/js-cookie";
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";

const PUBLIC_FILE = /\.(.*)$/;

export async function middleware(request: NextRequest) {
  //Ref:https://github.com/vercel/next.js/discussions/38615

  const auth = request.cookies.get(authCookieKey)?.value;
  const isTokenExpired = auth ? Date.now() > JSON.parse(auth).expired_at : true;

  const { pathname } = request.nextUrl;

  if (
    pathname.startsWith("/_next") || // exclude Next.js internals
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    PUBLIC_FILE.test(pathname) // exclude all files in the public folder
  )
    return NextResponse.next();

  if (pathname === "/") {
    if (auth) return NextResponse.redirect(new URL("/dashboard", request.url));

    return NextResponse.next();
  }

  if (!auth || isTokenExpired) {
    request.cookies.delete(authCookieKey);
    const response = NextResponse.redirect(new URL("/", request.url));
    response.cookies.delete(authCookieKey);
    return response;
  }

  return NextResponse.next();
}

`
this is my version of middleware which is working fine. how to handle the cookie auth in client side globally.
thanks in advance

**Comment by zargold** [comment]
Would you be able to accomplish the same with the matcher regex?

```JavaScript
export const config = {
  matcher: '/((?!api|_next|static|public|favicon.ico).*)'
}
```

**Reply by rob-mcgrail** [reply]
I see this regex a lot, and it has the major downside of not running on the homepage.

**Reply by chimit** [reply]
A stupid solution to cover the homepage:
```js
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!api|_next/static|_next/image|favicon.ico).*)',
    '/',
  ],
};
```

**Reply by rob-mcgrail** [reply]
Thanks for adding this - I actually worked it out a while ago but forgot that I'd left this comment. I feel like it could be clearer in the guides that you can just add multiple items to the matcher array (although now that I re-read the guide I see that's in the first example).

**Reply by chimit** [reply]
Yep, the main problem is that the guide doesn't cover the very basic case. @leerob

**Reply by lewi6** [reply]
am  using this and the middleware still runs on the home page 
<img width="582" alt="image" src="https://user-images.githubusercontent.com/55507100/221422064-4ce8c55b-681a-46a7-90cf-427ca1b90012.png">

**Reply by mamlzy** [reply]
my problem was middleware not triggered on `'/'` route. Thanks @chimit adding `'/'` in `mathcer` has fixed my issue

**Reply by Duchoang0201** [reply]
You save my life bro. Thank you so much!! @chimit

**Reply by itse4elhaam** [reply]
This should be the default behavior.

Also nextjs docs lacking this is astonishing.

**Reply by fo-fo** [reply]
Even the Next.js documentation at https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy#adding-a-nonce-with-middleware has that matcher which doesn't match the root `/` path 🫠

**Comment by mattschwarz** [comment]
I was banging my head against the wall for too long on this, so very glad this discussion took place. I submitted a PR to the Vercel example repo for Maintenance Pages using @icyJoseph fix: https://github.com/vercel/examples/pull/558

Appreciate any feedback/reviews. Definitely think this should be part of the Vercel example since within the context of creating something like a maintenance page, I don't think it's obvious that all paths are being captured. I imagine in most scenarios Vercel's example would break apps' maintenance pages.

**Comment by jimjamdev** [comment]
How can we ignore public images? I have a manifest.json and public/images/icons and I need to ignore the whole folder or it's not finding the icons. I have a bunch of language based redirects also. The first page load, shows the 404 until i refresh.

```
import type { NextRequest } from 'next/server';
import { NextResponse } from 'next/server';
import { match as matchLocale } from '@formatjs/intl-localematcher';
// import countries from '@/app/json/countries.json';
import { config as AppConfig } from '@/config';

const { language } = AppConfig;

/** Function to get the current browser language */

function getLocale(request: NextRequest): string | undefined {
	const acceptLanguage = request.headers.get('accept-language');
	const browserLanguages = acceptLanguage
		?.split(',')
		.map((lang) => lang.split(';')[0]) || ['en-US'];
	// @ts-expect-error locales are readonly
	const languages: string[] = language?.languages;
	return matchLocale(browserLanguages, languages, language?.defaultLanguage);
}

/** Middleware to redirect to the correct locale */

export function middleware(request: NextRequest) {
	const pathname = request.nextUrl.pathname;

	// // `/_next/` and `/api/` are ignored by the watcher, but we need to ignore files in `public` manually.
	// // If you have one
	/*if (
		[
			'/manifest.json',
			'/favicon.ico',
			// Your other files in `public`
		].includes(pathname)
	)
		return;*/

	const { geo, cookies } = request;
	const currentGeoCookie = cookies.get('geo');
	/**
	 * Set the GeoLocation cookie
	 * */
	if (!currentGeoCookie) {
		const response = NextResponse.next();
		response.cookies.set('geo', JSON.stringify(geo), {
			maxAge: 60 * 60 * 24 * 30,
			path: '/',
		});
		return response;
	}

	// Check if there is any supported locale in the pathname
	const pathnameIsMissingLocale = language?.languages.every(
		(locale) =>
			!pathname.startsWith(`/${locale}/`) && pathname !== `/${locale}`,
	);

	// Redirect if there is no locale
	if (pathnameIsMissingLocale) {
		const langCode = getLocale(request) || language?.defaultLanguage;

		// e.g. incoming request is /products
		// The new URL is now /en-US/products
		return NextResponse.redirect(
			new URL(`/${langCode}/${pathname}`, request.url),
		);
	}
}

export const config = {
	// Matcher ignoring `/_next/`, `/api/` '/images' and '/public' subpaths
	matcher: [
		'/((?!api|_next/static|_next/image|robots.txt|public|images|manifest.json|sw.js|favicon.ico|workbox-*).*)',
		'/'
	],
};

```

**Comment by jimjamdev** [comment]
Ok, turned out to be my cookies. Removed it and it all worked.

Guess there's still a bug here.

**Comment by dmudro** [comment]
I am having a similar issue: using the matcher with negative lookhead does not catch root route `/` in _Vercel_. works in localhost (webpack dev server) though.

adding a new `/` matcher causes a redirect loop in localhost but fixes Vercel. @jimjamdev sounds like you run into this too? I am using the same negative lookahead regex and removing cookies didnt help - I get stuck in a loop. locally only.

see https://github.com/vercel/next.js/issues/62078#issuecomment-1959590967
